File: forex_bot/api/websocket.py
"""
websocket.py
=============
Flask/flask-sock route wiring for the WebSocket endpoint. Connection
bookkeeping (channels, sequencing, replay, heartbeat, backpressure) is
delegated to api.websocket_manager.connection_manager - this file only
owns the transport (accepting the socket, reading/writing frames,
translating client JSON actions into connection_manager calls).

PUBLIC INTERFACE PRESERVED: api/app.py calls init_websocket(app) and
nothing else from this module. `registry` is kept as a module-level
alias of connection_manager so code written against the previous
ConnectionRegistry-based version keeps working unmodified.

Protocol (client -> server JSON):
  {"token": "<jwt>"}                                   - required first message (auth)
  {"action": "subscribe", "channel": "trades"}
  {"action": "unsubscribe", "channel": "trades"}
  {"action": "replay", "channel": "trades", "since_sequence": 41}
  {"action": "ack", "event_id": "..."}
  {"action": "pong"}                                   - heartbeat response (optional)

Server -> client JSON:
  {"event_id","sequence","event","channel","payload","occurred_at"}    - a broadcast EventDTO
  {"action": "ping", "server_time": ...}                                - heartbeat probe
  {"action": "replay_complete", "channel", "events_sent", "replay_incomplete"}
  {"action": "subscribed"/"unsubscribed", "channel": ...}
  {"error": "..."}
"""

from __future__ import annotations

import json
import logging
import time

from flask_sock import Sock

from api.auth import AuthError, decode_access_token
from api.websocket_manager import connection_manager
from core.event_bus import Events, bus

logger = logging.getLogger("api.websocket")

sock = Sock()

_AUTH_TIMEOUT_SECONDS = 10

_EVENT_CHANNEL_MAP: dict[str, str] = {
    Events.BOT_STARTED: "bot", Events.BOT_STOPPED: "bot", Events.BOT_PAUSED: "bot",
    Events.KILL_SWITCH_TRIGGERED: "bot",
    Events.TRADE_OPENED: "trades", Events.TRADE_CLOSED: "trades", Events.TRADE_REJECTED: "trades",
    Events.SL_HIT: "trades", Events.TP_HIT: "trades",
    Events.POSITION_BREAKEVEN: "trades", Events.POSITION_TRAILING_UPDATED: "trades",
    Events.POSITION_PARTIAL_CLOSED: "trades",
    Events.SIGNAL_GENERATED: "signals", Events.SIGNAL_SCORED: "signals", Events.SIGNAL_REJECTED: "signals",
    Events.MT5_CONNECTED: "health", Events.MT5_DISCONNECTED: "health", Events.MT5_RECONNECT_FAILED: "health",
    Events.DAILY_LOSS_LIMIT_HIT: "risk", Events.DRAWDOWN_LIMIT_HIT: "risk", Events.DRAWDOWN_WARNING: "risk",
    Events.MARGIN_WARNING: "risk",
    Events.CIRCUIT_BREAKER_TRIPPED: "health", Events.CIRCUIT_BREAKER_RECOVERED: "health",
    Events.CONFIG_RELOADED: "config", Events.CRITICAL_ERROR: "health",
    "ACCOUNT_TICK": "account", "MARKET_TICK": "market", "RISK_TICK": "risk",
    "HEALTH_TICK": "health", "METRICS_TICK": "metrics", "EXECUTION_TICK": "execution",
}

# Channel authorization: which roles can subscribe to each channel.
# viewer has no channel subscriptions by default (read-only dashboard only).
_CHANNEL_ROLES: dict[str, set[str]] = {
    "account": {"owner", "admin", "trader"},
    "bot": {"owner", "admin"},
    "trades": {"owner", "admin", "trader"},
    "signals": {"owner", "admin", "trader"},
    "risk": {"owner", "admin", "trader"},
    "health": {"owner", "admin", "trader", "viewer"},
    "config": {"owner", "admin"},
    "metrics": {"owner", "admin", "trader", "viewer"},
    "market": {"owner", "admin", "trader", "viewer"},
    "execution": {"owner", "admin"},
}


def _can_subscribe(role: str, channel: str) -> bool:
    allowed = _CHANNEL_ROLES.get(channel)
    if allowed is None:
        return role in {"owner", "admin"}
    return role in allowed


_subscribed_to_bus = False


def _on_domain_event(event) -> None:
    channel = _EVENT_CHANNEL_MAP.get(event.name)
    if channel is None:
        return
    connection_manager.broadcast(event.name, channel, event.payload)


def start_broadcasting() -> None:
    global _subscribed_to_bus
    if _subscribed_to_bus:
        return
    for event_name in _EVENT_CHANNEL_MAP:
        bus.subscribe(event_name, _on_domain_event)
    _subscribed_to_bus = True
    connection_manager.start_heartbeat()
    logger.info("WebSocket broadcaster subscribed to %d event types.", len(_EVENT_CHANNEL_MAP))


registry = connection_manager


def init_websocket(app) -> None:
    sock.init_app(app)
    start_broadcasting()

    @sock.route("/ws")
    def ws_handler(ws):
        auth_result = _authenticate(ws)
        if auth_result is None:
            try:
                ws.send(json.dumps({"error": "unauthorized"}))
            except Exception:
                pass
            return

        role = auth_result.get("role", "viewer")
        user_id = auth_result.get("user_id")
        conn_id = connection_manager.connect(ws, role, send_fn=ws.send, user_id=user_id)
        try:
            ws.send(json.dumps({
                "action": "connected", "conn_id": conn_id, "role": role,
                "current_sequence": connection_manager.sequencer.current_sequence(),
            }))
            _message_loop(ws, conn_id)
        finally:
            connection_manager.disconnect(conn_id)

    def _authenticate(ws):
        deadline = time.monotonic() + _AUTH_TIMEOUT_SECONDS
        while time.monotonic() < deadline:
            try:
                raw = ws.receive(timeout=_AUTH_TIMEOUT_SECONDS)
            except Exception:
                return None
            if raw is None:
                return None
            try:
                data = json.loads(raw)
            except json.JSONDecodeError:
                continue

            token = data.get("token", "")
            if not token:
                continue
            try:
                payload = decode_access_token(token)
                return payload
            except AuthError:
                return None
        return None

    def _message_loop(ws, conn_id):
        while True:
            raw = ws.receive()
            if raw is None:
                break

            connection_manager.touch(conn_id)

            try:
                data = json.loads(raw)
            except json.JSONDecodeError:
                _safe_send(ws, {"error": "Invalid JSON."})
                continue

            action = data.get("action")

            if action == "subscribe":
                channel = data.get("channel", "")
                if not channel:
                    _safe_send(ws, {"error": "subscribe requires a 'channel'."})
                    continue
                if not _can_subscribe(role, channel):
                    _safe_send(ws, {"error": f"Role '{role}' is not authorized for channel '{channel}'."})
                    continue
                connection_manager.subscribe(conn_id, channel)
                _safe_send(ws, {"action": "subscribed", "channel": channel})

            elif action == "unsubscribe":
                channel = data.get("channel", "")
                connection_manager.unsubscribe(conn_id, channel)
                _safe_send(ws, {"action": "unsubscribed", "channel": channel})

            elif action == "replay":
                channel = data.get("channel", "")
                since_sequence = int(data.get("since_sequence", 0))
                try:
                    connection_manager.send_replay(conn_id, channel, since_sequence)
                except Exception as exc:
                    _safe_send(ws, {"error": str(exc)})

            elif action == "ack":
                event_id = data.get("event_id", "")
                if event_id:
                    connection_manager.acknowledge(conn_id, event_id)

            elif action == "pong":
                pass

            else:
                _safe_send(ws, {"error": f"Unknown action '{action}'."})

    def _safe_send(ws, obj):
        try:
            ws.send(json.dumps(obj, default=str))
        except Exception:
            pass

File: dashboard/src/types/websocket.ts
// WebSocket event and protocol contracts shared by the dashboard's real-time
// client, hooks, stores, and components.

// -----------------------------------------------------------------------------
// Channels
// -----------------------------------------------------------------------------

export type WsChannel =
  | "bot"
  | "trades"
  | "signals"
  | "health"
  | "risk"
  | "config"
  | "account"
  | "market"
  | "metrics"
  | "execution";

// -----------------------------------------------------------------------------
// Autonomous decision contracts
// -----------------------------------------------------------------------------

export type AutonomousDecisionOutcome = "ALLOW" | "REJECT";

export type AutonomousDecisionDirection =
  | "BUY"
  | "SELL"
  | "HOLD";

export interface AutonomousDecisionIndicators {
  rsi: number | null;
  atr: number | null;
  atr_percent: number | null;
  adx: number | null;
  plus_di: number | null;
  minus_di: number | null;
}

export interface AutonomousDecisionPayload {
  outcome: AutonomousDecisionOutcome;
  allowed: boolean;
  symbol: string;
  timeframe: string;
  direction: AutonomousDecisionDirection;
  confidence: number;
  reason: string;
  checks: string[];
  indicators: AutonomousDecisionIndicators;
  evaluated_at: string;
}

// -----------------------------------------------------------------------------
// Event payload contracts
// -----------------------------------------------------------------------------

export interface BotLifecyclePayload {
  previous?: string;
  reason?: string;
}

export interface AccountTickPayload {
  balance?: number;
  equity?: number;
  margin?: number;
  free_margin?: number;
  margin_level?: number | null;
  currency?: string;
  leverage?: number | null;
  captured_at?: string;
}

export interface MarketTickPayload {
  symbol?: string;
  bid?: number;
  ask?: number;
  spread?: number;
  timestamp?: string;
}

export interface RiskTickPayload {
  daily_pnl?: number;
  daily_drawdown_pct?: number;
  max_drawdown_pct?: number;
  open_risk_pct?: number;
  exposure_pct?: number;
  risk_per_trade_pct?: number;
  open_positions?: number;
  margin_level?: number | null;
  captured_at?: string;
}

export interface WsEventMap {
  BOT_STARTED: BotLifecyclePayload;
  BOT_STOPPED: BotLifecyclePayload;
  BOT_PAUSED: BotLifecyclePayload;
  KILL_SWITCH_TRIGGERED: BotLifecyclePayload;

  TRADE_OPENED: Record<string, unknown>;
  TRADE_CLOSED: Record<string, unknown>;
  TRADE_REJECTED: Record<string, unknown>;
  SL_HIT: Record<string, unknown>;
  TP_HIT: Record<string, unknown>;
  POSITION_BREAKEVEN: Record<string, unknown>;
  POSITION_TRAILING_UPDATED: Record<string, unknown>;
  POSITION_PARTIAL_CLOSED: Record<string, unknown>;

  SIGNAL_GENERATED: Record<string, unknown>;
  SIGNAL_SCORED: Record<string, unknown>;
  SIGNAL_REJECTED: Record<string, unknown>;
  AUTONOMOUS_DECISION_EVENT: AutonomousDecisionPayload;

  MT5_CONNECTED: Record<string, unknown>;
  MT5_DISCONNECTED: Record<string, unknown>;
  MT5_RECONNECT_FAILED: Record<string, unknown>;

  DAILY_LOSS_LIMIT_HIT: Record<string, unknown>;
  DRAWDOWN_LIMIT_HIT: Record<string, unknown>;
  DRAWDOWN_WARNING: Record<string, unknown>;
  MARGIN_WARNING: Record<string, unknown>;

  CIRCUIT_BREAKER_TRIPPED: Record<string, unknown>;
  CIRCUIT_BREAKER_RECOVERED: Record<string, unknown>;
  CONFIG_RELOADED: Record<string, unknown>;
  CRITICAL_ERROR: Record<string, unknown>;

  ACCOUNT_TICK: AccountTickPayload;
  MARKET_TICK: MarketTickPayload;
  RISK_TICK: RiskTickPayload;
  HEALTH_TICK: Record<string, unknown>;
  METRICS_TICK: Record<string, unknown>;
  EXECUTION_TICK: Record<string, unknown>;
}

export type WsEventName = keyof WsEventMap;

// -----------------------------------------------------------------------------
// Event envelopes
// -----------------------------------------------------------------------------

export interface WsEvent<
  TEventName extends WsEventName = WsEventName,
> {
  event_id: string;
  sequence: number;
  event: TEventName;
  channel: WsChannel;
  payload: WsEventMap[TEventName];
  occurred_at: string;
}

export type AutonomousDecisionWsEvent =
  WsEvent<"AUTONOMOUS_DECISION_EVENT">;

// -----------------------------------------------------------------------------
// Server control messages
// -----------------------------------------------------------------------------

export interface WsConnected {
  action: "connected";
  conn_id: string;
  role: string;
  current_sequence: number;
}

export interface WsPing {
  action: "ping";
  server_time: number;
}

export interface WsReplayComplete {
  action: "replay_complete";
  channel: WsChannel;
  events_sent: number;
  replay_incomplete: boolean;
}

export interface WsSubscribed {
  action: "subscribed";
  channel: WsChannel;
}

export interface WsUnsubscribed {
  action: "unsubscribed";
  channel: WsChannel;
}

export interface WsError {
  error: string;
}

export type WsControlMessage =
  | WsConnected
  | WsPing
  | WsReplayComplete
  | WsSubscribed
  | WsUnsubscribed
  | WsError;

export type WsServerMessage =
  | WsEvent
  | WsControlMessage;

// -----------------------------------------------------------------------------
// Type guards
// -----------------------------------------------------------------------------

export function isWsEvent(
  message: WsServerMessage,
): message is WsEvent {
  return (
    "event" in message &&
    "channel" in message &&
    "payload" in message
  );
}

export function isAutonomousDecisionEvent(
  event: WsEvent,
): event is AutonomousDecisionWsEvent {
  return event.event === "AUTONOMOUS_DECISION_EVENT";
}